src/Internal/Service/PaymentProcessingService.php

```diff
@@ -8,19 +8,129 @@
+use WCPay\Internal\Payment\Storage;
+use WCPay\Internal\Payment\Response\ResponseInterface;
+use WCPay\Internal\Proxy\LegacyProxy;
+use WCPay\WooPay\WooPay_Utilities;
 
 /**
  * Payment Processing Service.
  */
 class PaymentProcessingService {
 	/**
-	 * Process payment.
+	 * Storage for payments.
+	 *
+	 * @var Storage
+	 */
+	private $storage;
+
+	/**
+	 * Legacy proxy.
+	 *
+	 * @var LegacyProxy
+	 */
+	private $legacy_proxy;
+
+	/**
+	 * Subscriptions service.
+	 *
+	 * @var WC_Payments_Subscription_Service
+	 */
+	private $subscriptions_service;
+
+	/**
+	 * WooPay utils.
+	 *
+	 * @var WooPay_Utilities
+	 */
+	private $woopay_utilities;
+
+	/**
+	 * Class constructor.
 	 *
-	 * @param  int $order_id Order ID provided by WooCommerce core.
+	 * @param Storage                          $storage               Payment storage.
+	 * @param LegacyProxy                      $legacy_proxy          Legacy proxy.
+	 * @param WC_Payments_Subscription_Service $subscriptions_service Subscriptions service.
+	 */
+	public function __construct(
+		Storage $storage,
+		LegacyProxy $legacy_proxy,
+		WC_Payments_Subscription_Service $subscriptions_service,
+		WooPay_Utilities $woopay_utilities
+	) {
+		$this->storage               = $storage;
+		$this->legacy_proxy          = $legacy_proxy;
+		$this->subscriptions_service = $subscriptions_service;
+		$this->woopay_utilities      = $woopay_utilities;
+	}
+
+	/**
+	 * Process payment.
 	 *
-	 * @throws Exception
+	 * @param int  $order_id       Order ID provided by WooCommerce core.
+	 * @param bool $manual_capture Whether the payment should be captured manually. Defaults to false.
+	 * @return ResponseInterface   The response from processing the payment.
 	 */
-	public function process_payment( int $order_id ) {
-		throw new Exception( 'Re-engineering payment process is in-progress. Sit tight, and wait more!' );
+	public function process_payment( int $order_id, bool $manual_capture = false ) {
+		$order = wc_get_order( $order_id ); // ToDo: This function should not be called directly, but through a service!
+
+		if ( ! $order instanceof WC_Order ) {
+			return new Failure( __( 'Processing payment failed: Order could not be loaded.', 'woocommerce-payments' ) );
+		}
+
+		try {
+			// Warning: Dev mode only! Will be deleted soon.
+			$this->storage->enable();
+			$this->storage->cleanup_order_payment( $order );
+
+			$post_data = [
+				'payment_method'                             => WC_Payment_Gateway_WCPay::GATEWAY_ID,
+				'wcpay-payment-method'                       => 'pm_XYZ',
+				'wc-woocommerce_payments-new-payment-method' => true,
+			];
+
+			// Initialize the request and payment.
+			$request = new PaymentRequest( $this->legacy_proxy, $post_data );
+			$payment = $this->storage->get_order_payment( $order );
+
+			// Setup flags.
+			if ( $request->should_save_payment_method() ) {
+				$payment->get_context()->add_flag( Flag::SAVE_PAYMENT_METHOD_TO_STORE() );
```

Comment:
So far, flags seem like the only thing that would need to be provided straight to the context. We could consider a proxy method within states, but I'm not certain of its usefulness.

Alternatively, those could be provided as a parameter to `process`, considering that (at least without UPE), they are only required for payment processing, and should not change after calling it.

Reply:
> Ok, I think I understand what the PaymentRequest is here - it's an object representing the request made to the client backend, right?

Yes.

> In this case, I think we should deserialize the request into the context, so it could be something like...

Most flags (except the ones for saving a payment method) do not originate from the request, but are rather provided to the gateway.

Also, we'd be adding a responsibility to the payment context, which seems to be outside its scope. That's keeping in mind that we decided that static methods should be pure, and `from_request` is far from pure.

On the other hand, some flags (if we keep them as such, which is worth IMO) might change between requests. In those cases, we should strive to reuse the same payment context, meaning that we cannot use flags to construct it, but should add them once the object is available.